tests/unit/layer/volumetric/precomputed/test_precomputed.py

```diff
@@ -102,3 +104,225 @@ def test_ts_set_voxel_set_voxel_offset_chunk_and_data(clear_caches, mocker):
+    [
+        # test shorthand
+        [[[16, 16, 40]], ["16_16_40"]],
+        # test multiples
+        [[[16, 16, 40], [32, 32, 40]], ["16_16_40", "32_32_40"]],
+        # test minimum spec of just `resolution`
+        [[{"resolution": [16, 16, 40]}], ["16_16_40"]],
+        # test mixed
+        [[{"resolution": [16, 16, 40]}, [32, 32, 40]], ["16_16_40", "32_32_40"]],
+    ],
+)
+def test_add_scale(clear_caches, tmpdir, in_scales, expects):
+    precomputed._write_info = _write_info_notmock
+    tmp_layer = make_tmp_layer(tmpdir)
+    info_spec = PrecomputedInfoSpec(
+        reference_path=tmp_layer,
+        add_scales=in_scales,
+    )
+    info_spec.update_info(tmp_layer, on_info_exists="overwrite")
+    scales = get_info(tmp_layer)["scales"]
+    for expect in expects:
+        assert expect in [e["key"] for e in scales]
+
+
+@pytest.mark.parametrize(
+    "mode, expects",
+    [
+        ["merge", ["4_4_40", "8_8_40", "16_16_40"]],
+        ["replace", ["16_16_40"]],
+    ],
+)
+def test_add_scale_modes(clear_caches, tmpdir, mode, expects):
+    precomputed._write_info = _write_info_notmock
+    tmp_layer = make_tmp_layer(tmpdir)
+    info_spec = PrecomputedInfoSpec(
+        reference_path=tmp_layer,
+        add_scales=[[16, 16, 40]],
+        add_scales_mode=mode,
+    )
+    info_spec.update_info(tmp_layer, on_info_exists="overwrite")
+    scales = get_info(tmp_layer)["scales"]
+    for expect in expects:
+        assert expect in [e["key"] for e in scales]
+
+
+def test_add_scale_modes_error(clear_caches, tmpdir):
+    precomputed._write_info = _write_info_notmock
+    tmp_layer = make_tmp_layer(tmpdir)
+    with pytest.raises(RuntimeError):
+        info_spec = PrecomputedInfoSpec(
+            reference_path=tmp_layer,
+            add_scales=[[16, 16, 40]],
+            add_scales_mode="append",
+        )
+        info_spec.update_info(tmp_layer, on_info_exists="overwrite")
+
+
+def test_add_scale_full_entry(clear_caches, tmpdir):
+    precomputed._write_info = _write_info_notmock
+    tmp_layer = make_tmp_layer(tmpdir)
+    info_spec = PrecomputedInfoSpec(
+        reference_path=tmp_layer,
+        add_scales=[
+            {
+                "key": "asdf8",
+                "resolution": [8, 8, 40],
+                "chunk_sizes": [[32, 32, 1]],
+                "encoding": "raw",
+                "size": [1000, 1000, 1000],
+                "voxel_offset": [1, 2, 3],
+            },
+            {
+                "resolution": [33, 33, 40],
+                "chunk_sizes": [[7, 8, 9]],
+                "encoding": "raw",
+                "size": [1000, 1000, 1000],
+                "voxel_offset": [4, 5, 6],
+            },
+        ],
+    )
+    info_spec.update_info(tmp_layer, on_info_exists="overwrite")
+    scales = get_info(tmp_layer)["scales"]
+    assert "asdf8" in [e["key"] for e in scales]
+    assert "33_33_40" in [e["key"] for e in scales]
+    for scale in scales:
+        if scale["key"] == "asdf8":
```

Comment:
Hmm, do we want to even allow random keys, or add a check? I'm not sure what the use case would be for arbitrary strings.

Reply:
I think `key` here is just a path for one particular scale? It could be useful for someone to have multiple scales of the same resolution, for example at different chunk sizes

Reply:
Oh, I hadn't thought of that. Yeah, that sounds like something pretty useful - the only issue is that to support that we would need to add support for getting things like chunk size and offsets in a way that's decoupled from the resolution (right now, `get/set_offset`, `get/set_chunk_size` all assume X_Y_Z format for the key, and the rest of the codebase just goes off of VolumetricIndex.)

Reply:
@dodamih Good points but the workflow need not integrate every weird use cases. For general usage, having decoupled `key` could be useful when it's not easy to rename folders (e.g., prefixes in buckets) and the user somehow wanted to have multiple versions of the same resolution. This use case is pretty weird though and you're probably right we should put a check to make sure `key` follows X_Y_Z format until there is demand otherwise.

Reply:
@supersergiy @dodamih Do we even support selecting chunk sizes if there are options?

Reply:
>could be useful for someone to have multiple scales of the same resolution, for example at different chunk sizes

That edge case is covered somewhere else in the info file: The `scale['chunk_sizes']` expects a list of `[sx,sy,sz]` chunk sizes. I believe the intention was to let Neuroglancer pick the chunking which best fits the currently visible cross-section. But never tested it so far.

Reply:
Thanks! For now I think I'll restrict `key` to follow `x_y_z` res to ensure best compatibility with the rest of the framework.